### ISSUE TITLE
Same User Accessing Different Environments 

### ISSUE THREAD

**Issue by psychemedia** [plasmabio/tljh-repo2docker#37]
I'm sure I had this working before, but on a recently deployment of the `tljh-repo2docker` recipe, I seem to have got locked into only being able to run the environment I first ran as a new user and not selecting any other environment.

That is, if I have created two different environments (`environmentA` and `environmentB`), if I log in as a new user and launch environmentA, then shutdown that server, logout, log back in again, select environmentB and launch that, it's actually environmentA that launches?

That is, it seems the user has to commit to one environment rather than stopping and starting different environments?

**Comment by pierrepo** [comment]
Hi @psychemedia 
Humm this seems weird :confused: 
I confirm that a given user should connect to a first env, shutdown this server and then reconnect to another env.

For instance, user `stu123` has access to 2 envs : `cours-python` and `cours-bash-test`.

![image](https://user-images.githubusercontent.com/733861/90130633-17472880-dd6b-11ea-86f2-15bf1f4835d1.png)

The animated gif below shows how the user will connect to the first one and then to the second:

![Peek 13-08-2020 13-47](https://user-images.githubusercontent.com/733861/90131015-b4a25c80-dd6b-11ea-944d-7bafe1e5dcd7.gif)

**Comment by psychemedia** [comment]
@pierrepo Yep, that's what I thought I should happen. But whatever I select second time round, I still launch into the same container as the one I launched first. And it seems to start up really quickly...

Two tweaks I made to the config compared to your installer:

1. I switched to the firstuseauthenticator:

```
# Full path if required: /opt/tljh/hub/bin/tljh-config 
sudo tljh-config set auth.type firstuseauthenticator.FirstUseAuthenticator
sudo tljh-config set auth.FirstUseAuthenticator.create_users True
sudo tljh-config reload
```

2. Set up a domain and added https as per [here](https://github.com/ouseful-demos/ou-rclub#specifyng-a-domian-name-and-enabling-https). I can let you have an admin account if that helps.

Demo server is at https://arms.ouseful.org/ for next few hours, but then it will be ripped down.

(I did wonder if my browser was cacheing something but I seem to get the same effect in different browsers.)

**Comment by psychemedia** [comment]
A UI note... If you stop the server, then you are presented with a *Start Server* button. My expectation is that that would (re)start the server you just stopped, rather than take you to the start server selection page. Tweaking the button text to *Select and start server* would make things clearer?

**Comment by pierrepo** [comment]
Hello @psychemedia 

>And it seems to start up really quickly...

That's probably a sign you're still in the same env.

Domain and https should not interfere since it's also what we did with [Plasma](https://docs.plasmabio.org/en/latest/install/https.html).

Unfortunately, I'm not familiar with `firstuseauthenticator`. But as far as I understand in the [README](https://github.com/jupyterhub/firstuseauthenticator#why-have-a-password-db-and-not-use-pam-), `firstuseauthenticator` does not use the classical PAM Unix. This could explain the issue. Maybe @jtpio could have thought on this (but he's not yet available).

> Tweaking the button text to Select and start server would make things clearer?

Thanks for the suggestion. I'll make an issue for this :+1:

**Comment by psychemedia** [comment]
I did wonder if the way the users were handled might be the issue. I've had to destroy the server for now but I'll try to explore a bit more using different authenticators/user creation methods next week to see if I can recreate the issue.

**Comment by jtpio** [comment]
The authenticator should be agnostic to the rest of the configuration.

Meaning that Plasma uses `PAMAuthenticator` but the `firstuseauthenticator` should also work fine with `tljh-repo2docker`.

> 1. I switched to the firstuseauthenticator:

> ```
> # Full path if required: /opt/tljh/hub/bin/tljh-config 
> sudo tljh-config set auth.type firstuseauthenticator.FirstUseAuthenticator
> sudo tljh-config set auth.FirstUseAuthenticator.create_users True
> sudo tljh-config reload
> ```

@psychemedia is that the only tweak to the JupyterHub config?

**Comment by psychemedia** [comment]
I think I'd also set a certificate for https/ssl.

This is the (manual) script I used.

--tony

**Comment by jtpio** [comment]
Thanks @psychemedia for sharing this :+1: 

It looks like HTTPS shouldn't have any effect on this behavior.

Don't hesitate to post here if you spin up a new VM with this setup. Otherwise it should also be possible to [test locally](https://github.com/plasmabio/tljh-repo2docker/blob/master/CONTRIBUTING.md) and changing the [test config](https://github.com/plasmabio/tljh-repo2docker/blob/master/jupyterhub_config.py) to use the other authenticator.
